### PR TITLE
Ensure cache consistency by tracking registerred caches and deserializing them after file changes

### PR DESCRIPTION
**Pull request by JPWilli**

### CODE REVIEW

File: src/Microsoft.Identity.Client.Extensions.Msal/InternalsVisibleTo.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.Reflection;
```
Comment (JPWilli) [inline src/Microsoft.Identity.Client.Extensions.Msal/InternalsVisibleTo.cs]: Left behind from testing. Will revert.

File: src/Microsoft.Identity.Client.Extensions.Msal/Microsoft.Identity.Client.Extensions.Msal.csproj [hunk 2]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
 
     <!--This should be passed from the VSTS build-->
-    <ClientSemVer Condition="'$(ClientSemVer)' == ''">1.0.0-localbuild</ClientSemVer>
+    <ClientSemVer Condition="'$(ClientSemVer)' == ''">1.0.10-localbuild</ClientSemVer>
```
Comment (JPWilli) [inline src/Microsoft.Identity.Client.Extensions.Msal/Microsoft.Identity.Client.Extensions.Msal.csproj]: Left behind from testing, will revert.

File: src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs [hunk 3]
```diff
@@ -187,31 +198,47 @@ public static string UserRootDirectory
...
         /// <param name="tokenCache">Token Cache</param>
         public void RegisterCache(ITokenCache tokenCache)
         {
-            lock (_lockObject)
+            // OK, we have two nested locks here. We need to maintain a clear ordering to avoid deadlocks.
+            // 1. Use the CrossPlatLock which is respected by all processes and is used around all cache accesses.
+            // 2. Use _lockObject which is used in UnregisterCache, and is needed for all accesses of _registerredCaches.
+            //
+            // Here specifically, we don't want to set this.CacheLock because we're done with the lock by the end of the method.
+            using (var crossPlatLock =  CreateCrossPlatLock(_storageCreationProperties))
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs]: nit: remove extra space

File: src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs [hunk 4]
```diff
@@ -221,10 +248,36 @@ public void RegisterCache(ITokenCache tokenCache)
                     }
                 }
 
+                _registerredCaches[tokenCache] = _store.LastWriteTime;
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs]: typo:  _registeredCaches

File: src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs [hunk 5]
```diff
@@ -60,15 +60,26 @@ public class MsalCacheHelper
...
         /// <summary>
         /// Allows clients to listen for cache updates originating from disk.
         /// </summary>
         public event EventHandler<CacheChangedEventArgs> CacheChanged;
 
+        /// <summary>
+        /// Contains a reference to all caches currently registerred to synchronize with this MsalCacheHelper, along with
+        /// timestamp of the cache file the last time they deserialized.
+        /// </summary>
+        private readonly Dictionary<ITokenCache, DateTimeOffset> _registerredCaches = new Dictionary<ITokenCache, DateTimeOffset>();
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs]: Will we get a reasonable "GetHashCode" implementation off of the interface from .NET to ensure these don't collide?

File: src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs [hunk 6]
```diff
@@ -221,10 +248,36 @@ public void RegisterCache(ITokenCache tokenCache)
...
         }
 
+
+        /// <summary>
+        /// Unregisters a token cache so it no longer synchronizes with on disk storage.
+        /// </summary>
+        /// <param name="tokenCache"></param>
+        public void UnregisterCache(ITokenCache tokenCache)
+        {
+            lock (_lockObject)
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs]: null check for against tokenCache

File: src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs [hunk 7]
```diff
@@ -187,31 +198,47 @@ public static string UserRootDirectory
...
         /// <summary>
-        /// Registers this object to receive events when the provided token cache changes, in order
-        /// to serialize to disk.
+        /// Registers a token cache to synchronize with on disk storage.
         /// </summary>
         /// <param name="tokenCache">Token Cache</param>
         public void RegisterCache(ITokenCache tokenCache)
         {
-            lock (_lockObject)
+            // OK, we have two nested locks here. We need to maintain a clear ordering to avoid deadlocks.
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs]: null check against tokenCache

File: src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs [hunk 8]
```diff
@@ -60,15 +60,26 @@ public class MsalCacheHelper
...
 
+        /// <summary>
+        /// Contains a reference to all caches currently registerred to synchronize with this MsalCacheHelper, along with
+        /// timestamp of the cache file the last time they deserialized.
+        /// </summary>
+        private readonly Dictionary<ITokenCache, DateTimeOffset> _registerredCaches = new Dictionary<ITokenCache, DateTimeOffset>();
+
         /// <summary>
         /// Creates a new instance of <see cref="MsalCacheHelper"/>.
         /// </summary>
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs]: This method should also do a null check against storageCreationProperties

**Pull request closed without merging**